src/state.rs

```diff
@@ -341,9 +372,39 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
+                    .is_none()
```

Comment:
what about else branch? seems you donot handle it

Reply:
> what about else branch? seems you donot handle it

Hi @hzxuzhonghu, okay, there should be waiting for the DNS resolving task because I added the remove element logic.